### PR TITLE
Fix #1963 and #2010: Prevent GCC to remove the `constructor` prop.

### LINKED ISSUE #1963: @ScalaJSDefined classes constructor pointing to wrong ctor in base/sub classes
Its happening in `fullOptJS` mode only 

**fullOptJS** : 

screen shot 2015-10-09 at 3 36 04 am

**fastOptJS** : 

screen shot 2015-10-09 at 3 34 59 am

Steps to reproduce : 

js code
```js

<script>
    "use strict";

    function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

    var BaseMutation = (function () {
        function BaseMutation() {
            _classCallCheck(this, BaseMutation);
        }

        BaseMutation.prototype.checkThis = function checkThis() {
            console.log(this.constructor);
            console.log(this.constructor.fragments);
        };

        return BaseMutation;
    })();
</script>
```

scala code : 

```scala
@js.native
class BaseMutation extends js.Object {
  def checkThis(): Unit = js.native
}

@ScalaJSDefined
class ScalaMutation extends BaseMutation

new ScalaMutation().checkThis() // check console this is referring to different constructors in two output modes

```

more context : https://github.com/scala-js/scala-js/issues/1902#issuecomment-146702496

Currently this is a blocker for me , i can't run relay apps in production mode! , I am happy to investigate/work on this with some pointers/guidelines ;)

### PR DESCRIPTION
**Pull request by sjrd**
Under some circumstances, the Closure compiler would remove the
reassignment of the `constructor` property on the prototype of
classes. This removal caused the issues seen in #1963 and #2010.

This commit fixes this issue by assigning it as StringLiteral
instead of an Ident.

The tests for this fix are put in javalibExTestSuite instead of
testSuite because, if put in testSuite, they "fail to fail",
probably due to mass effects of the testSuite codebase, which
cause Closure not to exhibit the bug.

### CODE REVIEW

**Comment by sjrd** [comment]
Review by @nicolasstucki

File: javalib-ex-test-suite/src/test/scala/scala/scalajs/testsuite/javalib/lang/ObjectTestEx.scala [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ *  codebase than testSuite to be meaningful.
+ *
+ *  If moved to testSuite, those tests "fail to fail" due to mass effects
+ *  produced by the immensity of the testSuite codebase.
+ */
+object ObjectTestEx extends JasmineTest {
+
+  describe("java.lang.Object ex") {
+
+    it("clone()") {
```
Comment (gzm0) [inline javalib-ex-test-suite/src/test/scala/scala/scalajs/testsuite/javalib/lang/ObjectTestEx.scala]: Reference the clone bug?

File: tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala [hunk 2]
```diff
@@ -245,7 +245,7 @@ final class ScalaJSClassEmitter private (
       js.Block(
           inheritedCtorDef,
           js.Assign(typeVar.prototype, js.New(inheritedCtorRef, Nil)),
-          genAddToPrototype(className, js.Ident("constructor"), typeVar)
+          genAddToPrototype(className, js.StringLiteral("constructor"), typeVar)
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala]: We also rely on the `constructor` property as dot select in the [`scalajsenv`][1] (for `TypedArray` backing of `scala.Array`). IIUC this case could break as well. Maybe we should add `constructor` to the externs?

[1]: https://github.com/scala-js/scala-js/blob/master/tools/scalajsenv.js#L851

Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala]: It *is* already in the externs. The reference in `scalajsenv` is not renamed. It was just the reassignments on the prototype that "disappeared" for some reason.

Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala]: Oh... OK. Strange. So this is more of a workaround than an actual bugfix in Scala.js

**Comment by sjrd** [comment]
Severe impact on checksizes: 7 KB non gzipped and 2 KB gzipped on fullOpt versions. But well, it's not like we have a choice.

File: tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala (hunk 2 above)
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala]: Yes, I think so.

**Comment by nicolasstucki** [comment]
LGTM.

**Comment by gzm0** [comment]
Ouch.. :(

**Pull request merged**